File: frontend/projects/valtimo/case-management/src/lib/services/startable-item-management.service.ts
/*
 * Copyright 2015-2026 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Injectable} from '@angular/core';
import {NGXLogger} from 'ngx-logger';
import {CaseManagementParams, CaseProcessDefinitionResponseDto} from '@valtimo/shared';
import {runAfterCarbonModalClosed} from '@valtimo/components';
import {BehaviorSubject, Observable, of, Subject} from 'rxjs';
import {catchError, filter, map, switchMap, take} from 'rxjs/operators';
import {
  BuildingBlockItemProperties,
  CreateStartableItemRequest,
  ManagementStartableItem,
  StartableItemOrderEntry,
  StartableItemType,
  UpdateStartableItemOrderRequest,
} from '../models';
import {StartableItemApiService} from './startable-item-api.service';

@Injectable()
export class StartableItemManagementService {
  private readonly _params$ = new BehaviorSubject<CaseManagementParams | null>(null);

  private readonly _items$ = new BehaviorSubject<ManagementStartableItem[]>([]);
  public readonly items$ = this._items$.asObservable();

  private readonly _loading$ = new BehaviorSubject<boolean>(false);
  public readonly loading$ = this._loading$.asObservable();

  private readonly _showModal$ = new BehaviorSubject<boolean>(false);
  public readonly showModal$ = this._showModal$.asObservable();

  private readonly _editingItem$ = new BehaviorSubject<ManagementStartableItem | null>(null);
  public readonly editingItem$ = this._editingItem$.asObservable();

  private readonly _showDeleteModal$ = new BehaviorSubject<boolean>(false);
  public readonly showDeleteModal$ = this._showDeleteModal$.asObservable();

  private readonly _itemToDelete$ = new BehaviorSubject<ManagementStartableItem | null>(null);
  public readonly itemToDelete$ = this._itemToDelete$.asObservable();

  private readonly _reorderComplete$ = new Subject<void>();
  public readonly reorderComplete$ = this._reorderComplete$.asObservable();

  public readonly usedProcessDefinitionIds$: Observable<string[]> = this._items$.pipe(
    map(items =>
      items
        .filter(item => item.type === StartableItemType.PROCESS && !!item.processDefinitionId)
        .map(item => item.processDefinitionId!)
    )
  );

  public readonly usedBuildingBlockKeys$: Observable<string[]> = this._items$.pipe(
    map(items =>
      items.filter(item => item.type === StartableItemType.BUILDING_BLOCK).map(item => item.key)
    )
  );

  public readonly linkedProcessDefinitions$: Observable<CaseProcessDefinitionResponseDto[]> =
    this._params$.pipe(
      filter((params): params is CaseManagementParams => params !== null),
      switchMap(params =>
        this.startableItemApiService.getLinkedProcessDefinitions(params)
      )
    );

  constructor(
    private readonly startableItemApiService: StartableItemApiService,
    private readonly logger: NGXLogger
  ) {}

  public setParams(params: CaseManagementParams): void {
    this._params$.next(params);
  }

  public getParams(): CaseManagementParams {
    const params = this._params$.value;
    if (!params) {
      this.logger.error('Params not initialized. Call setParams() first.');
    }
    return params!;
  }

  public loadItems(): void {
    this._loading$.next(true);
    this.startableItemApiService
      .getItems(this.getParams())
      .pipe(take(1))
      .subscribe({
        next: (items: ManagementStartableItem[]) => {
          this._items$.next(items);
          this._loading$.next(false);
        },
        error: error => {
          console.error(error);
          this._loading$.next(false);
        },
      });
  }

  public showCreateModal(): void {
    this._editingItem$.next(null);
    this._showModal$.next(true);
  }

  public showEditModal(item: ManagementStartableItem): void {
    this._editingItem$.next(item);
    this._showModal$.next(true);
  }

  public hideModal(): void {
    this._showModal$.next(false);
    runAfterCarbonModalClosed(() => this._editingItem$.next(null));
  }

  public showDeleteConfirmation(item: ManagementStartableItem): void {
    this._itemToDelete$.next(item);
    this._showDeleteModal$.next(true);
  }

  public hideDeleteModal(): void {
    this._showDeleteModal$.next(false);
    this._itemToDelete$.next(null);
  }

  public createItem(request: CreateStartableItemRequest): Observable<ManagementStartableItem> {
    return this.startableItemApiService.createItem(this.getParams(), request);
  }

  public updateItem(
    key: string,
    versionTag: string | null,
    request: CreateStartableItemRequest
  ): Observable<ManagementStartableItem> {
    return this.startableItemApiService.updateItem(
      this.getParams(),
      key,
      versionTag,
      request
    );
  }

  public getItemProperties(
    itemKey: string,
    versionTag: string | null,
    type: StartableItemType
  ): Observable<BuildingBlockItemProperties> {
    return this.startableItemApiService.getItemProperties(
      this.getParams(),
      itemKey,
      versionTag,
      type
    );
  }

  public deleteItem(item: ManagementStartableItem): Observable<void> {
    return this.startableItemApiService.deleteItem(
      this.getParams(),
      item.key,
      item.versionTag
    );
  }

  public updateOrder(items: ManagementStartableItem[]): void {
    this._items$.next(items);

    const request: UpdateStartableItemOrderRequest = {
      items: items.map(
        (item, index): StartableItemOrderEntry => ({
          key: item.key,
          type: item.type,
          versionTag: item.versionTag,
          sortOrder: index,
        })
      ),
    };

    this.startableItemApiService
      .updateOrder(this.getParams(), request)
      .pipe(
        take(1),
        catchError(() => {
          this.loadItems();
          return of([]);
        })
      )
      .subscribe({
        next: (updatedItems: ManagementStartableItem[]) => {
          if (updatedItems.length) {
            this._items$.next(updatedItems);
          }
          this._reorderComplete$.next();
        },
      });
  }
}

File: backend/case/src/main/kotlin/com/ritense/case/service/StartableItemManagementService.kt
/*
 * Copyright 2015-2026 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ritense.case.service

import com.fasterxml.jackson.databind.JsonNode
import com.ritense.case.domain.StartableItem
import com.ritense.case.domain.StartableItemId
import com.ritense.case.repository.StartableItemRepository
import com.ritense.case.web.rest.dto.ManagementStartableItemDto
import com.ritense.case.web.rest.dto.StartableItemDto
import com.ritense.case.web.rest.dto.StartableItemOrderEntry
import com.ritense.case.web.rest.dto.StartableItemType
import com.ritense.valtimo.contract.annotation.SkipComponentScan
import com.ritense.valtimo.contract.case_.CaseDefinitionId
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@SkipComponentScan
@Service
class StartableItemManagementService(
    private val startableItemProviders: List<StartableItemProvider>,
    private val startableItemRepository: StartableItemRepository,
) {

    @Transactional(readOnly = true)
    fun getStartableItems(caseDefinitionId: CaseDefinitionId): List<ManagementStartableItemDto> {
        val allItems = startableItemProviders.flatMap { provider ->
            provider.getStartableItems(caseDefinitionId)
        }

        val sortOrderMap = startableItemRepository
            .findAllByIdCaseDefinitionId(caseDefinitionId)
            .associate { Triple(it.id.itemKey, it.id.itemType, it.id.itemVersionTag) to it.sortOrder }

        return allItems
            .map { item ->
                ManagementStartableItemDto(
                    type = item.type,
                    name = item.name,
                    key = item.key,
                    versionTag = item.versionTag,
                    processDefinitionId = item.processDefinitionId,
                    sortOrder = sortOrderMap[Triple(item.key, item.type, item.versionTag.orEmpty())]
                )
            }
            .sortedWith(compareBy(
                { it.sortOrder ?: Int.MAX_VALUE },
                { it.name ?: it.key }
            ))
    }

    @Transactional
    fun createItem(
        caseDefinitionId: CaseDefinitionId,
        type: StartableItemType,
        properties: JsonNode
    ): StartableItemDto {
        val provider = startableItemProviders.find { it.type == type }
            ?: throw UnsupportedOperationException("No provider found for type: $type")
        val item = provider.createItem(caseDefinitionId, properties)

        val maxSortOrder = startableItemRepository
            .findAllByIdCaseDefinitionId(caseDefinitionId)
            .maxOfOrNull { it.sortOrder } ?: -1

        startableItemRepository.save(
            StartableItem(
                id = StartableItemId(caseDefinitionId, item.key, item.type, item.versionTag.orEmpty()),
                sortOrder = maxSortOrder + 1
            )
        )

        return item
    }

    @Transactional
    fun updateItem(
        caseDefinitionId: CaseDefinitionId,
        oldItemKey: String,
        oldVersionTag: String?,
        newType: StartableItemType,
        newProperties: JsonNode
    ): StartableItemDto {
        val existingItems = startableItemProviders.flatMap { it.getStartableItems(caseDefinitionId) }
        val oldItem = existingItems.find { it.key == oldItemKey && it.versionTag == oldVersionTag }
            ?: throw NoSuchElementException("Startable item not found: $oldItemKey:$oldVersionTag")

        val oldSortOrder = startableItemRepository
            .findAllByIdCaseDefinitionId(caseDefinitionId)
            .find { it.id.itemKey == oldItemKey && it.id.itemType == oldItem.type && it.id.itemVersionTag == oldVersionTag.orEmpty() }
            ?.sortOrder

        if (oldItem.type == newType) {
            val provider = startableItemProviders.find { it.type == newType }
                ?: throw UnsupportedOperationException("No provider found for type: $newType")
            return provider.updateItem(caseDefinitionId, oldItemKey, oldVersionTag, newProperties)
        }

        val oldProvider = startableItemProviders.find { it.type == oldItem.type }
            ?: throw UnsupportedOperationException("No provider found for type: ${oldItem.type}")
        oldProvider.deleteItem(caseDefinitionId, oldItemKey, oldVersionTag)
        startableItemRepository.deleteById(StartableItemId(caseDefinitionId, oldItemKey, oldItem.type, oldVersionTag.orEmpty()))

        val newProvider = startableItemProviders.find { it.type == newType }
            ?: throw UnsupportedOperationException("No provider found for type: $newType")
        val newItem = newProvider.createItem(caseDefinitionId, newProperties)

        val sortOrder = oldSortOrder ?: (startableItemRepository
            .findAllByIdCaseDefinitionId(caseDefinitionId)
            .maxOfOrNull { it.sortOrder }?.plus(1) ?: 0)

        startableItemRepository.save(
            StartableItem(
                id = StartableItemId(caseDefinitionId, newItem.key, newItem.type, newItem.versionTag.orEmpty()),
                sortOrder = sortOrder
            )
        )

        return newItem
    }

    @Transactional
    fun deleteItem(
        caseDefinitionId: CaseDefinitionId,
        itemKey: String,
        versionTag: String?
    ) {
        val existingItems = startableItemProviders.flatMap { it.getStartableItems(caseDefinitionId) }
        val item = existingItems.find { it.key == itemKey && it.versionTag == versionTag }
            ?: throw NoSuchElementException("Startable item not found: $itemKey:$versionTag")

        val provider = startableItemProviders.find { it.type == item.type }
            ?: throw UnsupportedOperationException("No provider found for type: ${item.type}")

        provider.deleteItem(caseDefinitionId, itemKey, versionTag)

        startableItemRepository.deleteById(StartableItemId(caseDefinitionId, itemKey, item.type, versionTag.orEmpty()))
    }

    @Transactional(readOnly = true)
    fun getItemProperties(
        caseDefinitionId: CaseDefinitionId,
        itemKey: String,
        versionTag: String?,
        type: StartableItemType
    ): JsonNode {
        val provider = startableItemProviders.find { it.type == type }
            ?: throw UnsupportedOperationException("No provider found for type: $type")
        return provider.getItemProperties(caseDefinitionId, itemKey, versionTag)
            ?: throw UnsupportedOperationException("Provider for type $type does not support item properties")
    }

    @Transactional
    fun updateOrder(
        caseDefinitionId: CaseDefinitionId,
        items: List<StartableItemOrderEntry>
    ): List<ManagementStartableItemDto> {
        val existingItems = startableItemProviders.flatMap { it.getStartableItems(caseDefinitionId) }
        val existingItemsByIdentity = existingItems.associateBy { Triple(it.key, it.type, it.versionTag.orEmpty()) }

        val entities = items.mapNotNull { entry ->
            existingItemsByIdentity[Triple(entry.key, entry.type, entry.versionTag.orEmpty())] ?: return@mapNotNull null
            StartableItem(
                id = StartableItemId(caseDefinitionId, entry.key, entry.type, entry.versionTag.orEmpty()),
                sortOrder = entry.sortOrder
            )
        }

        startableItemRepository.deleteAllByIdCaseDefinitionId(caseDefinitionId)
        startableItemRepository.saveAll(entities)

        return getStartableItems(caseDefinitionId)
    }
}
